lib/http3/common.c

```diff
@@ -1368,6 +1387,18 @@ uint64_t h2o_http3_decode_h3_datagram(h2o_iovec_t *payload, const void *_src, si
     uint64_t flow_id;
 
     if ((flow_id = ptls_decode_quicint(&src, end)) != UINT64_MAX)
+    {
+        //After the quarter stream id you get the context-id which for UDP should be 0 and we point the payload to the byte after that
+        // (see comments to RFCs in the h2o_http3_send_h3_datagrams function above)
+        if (*src == 0) {
+            src++;
+        } else {
+            // H3 Context-ID not 0
+            return UINT64_MAX;
+        }
         *payload = h2o_iovec_init(src, end - src);
+    }
+    // convert from quarter stream id
+    flow_id = flow_id << 2;
     return flow_id;
```

Comment:
unlike `h2o_http3_send_h3_datagrams()`, `h2o_http3_decode_h3_datagram()` does not have `h2o_http3_conn_t *conn` as a parameter, suggest adding it and passing it from `lib/http3/server.c::datagram_frame_receive_cb()` and `lib/common/http3client.c::on_receive_datagram_frame()`.

Then the rfc logic could be guarded in blocks.


```suggestion
     if ((flow_id = ptls_decode_quicint(&src, end)) != UINT64_MAX) {
        if (conn->peer_settings.h3_datagram_rfc) {
            // After the quarter stream id you get the context-id which for UDP should be 0 and we point the payload to the byte
            // after that see comments to RFCs in the h2o_http3_send_h3_datagrams function above)
            if (*src == 0) {
                src++;
            } else {
                // H3 Context-ID not 0
                return UINT64_MAX;
            }
        }
        *payload = h2o_iovec_init(src, end - src);
    }
    if (conn->peer_settings.h3_datagram_rfc) {
        // convert from quarter stream id
        flow_id = flow_id << 2;
    }
    return flow_id;

```

That is of course unless this logic is moved within the connect handler.